packages/fast-tooling-react/src/css-property-editor/property-editor.tsx

```diff
@@ -1,237 +1,365 @@
+        Object.keys(this.editData).forEach((key: string, index: number) => {
+            if (index === this.currentEditRowIndex) {
+                newData[
+                    this.state.currentEditRowUncommittedCSSPropertyKey
+                ] = this.editData[key];
+            } else {
+                newData[key] = this.editData[key];
+            }
+        });
+
+        this.editData = newData;
         this.setState({
-            key: e.target.value,
+            currentEditRowUncommittedCSSPropertyKey: null,
         });
+
+        this.handleCSSUpdate(newData);
     };
 
-    private handleNewPropertyValueChange = (
-        e: React.ChangeEvent<HTMLInputElement>
-    ): void => {
+    /**
+     * Row gained focus
+     */
+    private handleRowFocus = (rowKey: string, rowIndex: number): void => {
+        if (this.currentEditRowIndex !== rowIndex) {
+            this.currentEditRowReactKey = null;
+            this.currentEditRowIndex = rowIndex;
+        }
+    };
+
+    /**
+     * Key input has lost focus
+     */
+    private handleCommitKeyEdit = (rowKey: string, rowIndex: number): void => {
+        if (this.state.currentEditRowUncommittedCSSPropertyKey !== null) {
+            if (this.state.currentEditRowUncommittedCSSPropertyKey === "") {
+                this.deleteRow(rowIndex);
+            } else {
+                this.commitCSSKeyEdit();
+            }
+        } else if (rowKey === "") {
+            this.deleteRow(rowIndex);
+        }
+    };
+
+    /**
+     * Row has lost focus
+     */
+    private handleRowBlur = (rowKey: string, rowIndex: number): void => {
+        if (this.currentEditRowIndex !== -1 && this.editData[rowKey] === "") {
+            this.deleteRow(rowIndex);
+        }
+        this.currentEditRowReactKey = null;
         this.setState({
-            value: e.target.value,
+            currentEditRowUncommittedCSSPropertyKey: null,
         });
     };
 
-    private handleNewPropertyValueKeyDown = (
-        e: React.KeyboardEvent<HTMLInputElement>
-    ): void => {
-        if (e.keyCode === KeyCodes.tab) {
-            if (this.state.key !== "" && this.state.value !== "") {
-                e.preventDefault();
+    /**
+     * Clicks on a row outside the inputs add a row following that row
+     */
+    private handleClickToInsert = (rowKey: string, rowIndex: number): void => {
+        this.createRow(rowIndex + 1);
+    };
+
+    /**
+     * Add a row at the end of the list when there are clicks outside of an existing row
+     */
+    private handleClick = (e: React.MouseEvent<HTMLDivElement>): void => {
+        if (e.defaultPrevented) {
+            return;
+        }
+        this.createRow(Object.keys(this.editData).length);
+    };
+
+    /**
+     * Component got focus without any data rows, so add an empty one
+     */
+    private handleEmptyFocus = (): void => {
+        this.createRow(0);
+    };
+
+    /**
+     * Create a new row at the insertion index
```

Comment:
can you add a bit more context here as it took me a while to understand it, something like:

```
/**
 * Create a new row at an insertion index
 *
 * Rows before will retain their index while rows that
 * are after the insertion index will refer to the previous index
 * to account for the newly inserted row
 */
```

Reply:
A simple "Retains keys for existing elements" is really all that is going on here.  Added that.